service/history/queuev2/pending_task_tracker.go

```diff
@@ -0,0 +1,85 @@
+// The MIT License (MIT)
+
+// Copyright (c) 2017-2020 Uber Technologies Inc.
+
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in all
+// copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+// SOFTWARE.
+
+package queuev2
+
+import (
+	"github.com/uber/cadence/common/persistence"
+	ctask "github.com/uber/cadence/common/task"
+	"github.com/uber/cadence/service/history/task"
+)
+
+type (
+	PendingTaskTracker interface {
+		AddTask(task.Task)
+		PruneAckedTasks()
+		GetMinimumTaskKey() (persistence.HistoryTaskKey, bool)
+		GetTasks() map[persistence.HistoryTaskKey]task.Task
+	}
+
+	pendingTaskTrackerImpl struct {
+		taskMap    map[persistence.HistoryTaskKey]task.Task
+		minTaskKey persistence.HistoryTaskKey
+	}
+)
+
+func NewPendingTaskTracker() PendingTaskTracker {
+	return &pendingTaskTrackerImpl{
+		taskMap: make(map[persistence.HistoryTaskKey]task.Task),
+	}
+}
+
+func (t *pendingTaskTrackerImpl) AddTask(task task.Task) {
+	if len(t.taskMap) == 0 {
+		t.minTaskKey = task.GetTaskKey()
+	} else if t.minTaskKey.Compare(task.GetTaskKey()) > 0 {
+		t.minTaskKey = task.GetTaskKey()
+	}
+
+	t.taskMap[task.GetTaskKey()] = task
+}
+
+func (t *pendingTaskTrackerImpl) GetMinimumTaskKey() (persistence.HistoryTaskKey, bool) {
+	if len(t.taskMap) == 0 {
+		return persistence.MaxHistoryTaskKey, false
+	}
+	return t.minTaskKey, true
+}
+
+func (t *pendingTaskTrackerImpl) GetTasks() map[persistence.HistoryTaskKey]task.Task {
+	return t.taskMap
```

Comment:
Do we expect callers to mutate this map? If so that will make minTaskKey obsolete. 
To avoid data copies and stay on the performant side, we can keep this as is but please document the desired behavior in the code comments to prevent misuse.